### PR TITLE
Servant MultiVerb

### PR DESCRIPTION
**Pull request by pcapriotti**
This introduces a new `Verb`-like type for Servant endpoints: `MultiVerb`. The `MultiVerb` type can be used as an alternative to `UVerb`, and it provides approximately the same functionality as the latter, with a few differences:

- The handler return type is provided by the user, and does not have to be a union. In fact, `MultiVerb` can subsume `Verb`.
- It is possible to associate arbitrary metadata to responses without affecting the return type.
- It is designed to be extensible with other kinds of responses.

This PR only adds the `MultiVerb` type and uses it for the two user HEAD endpoints (qualified and unqualified). The followup PR #1657 applies `MultiVerb` to a few other endpoints and makes `ErrorDescription` types work as `MultiVerb` responses.

## Checklist

 - [x] Title of this PR explains the impact of the change.
 - [x] The description provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] The CHANGELOG.md file in the *Unreleased* section has been updated to explain the change which will be included in the release notes.
 - [x] Added documentation.
 - [x] Implemented Swagger generation
 - [x] Implemented header support
 - [x] Implemented Client support

### CODE REVIEW

**Review by mdimjasevic — APPROVED** [review]
This looks good! I left a few minor comments.

File: libs/wire-api/src/Wire/API/Routes/MultiVerb.hs [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+  UnrenderSuccess x >>= f = f x
+
+instance Alternative UnrenderResult where
+  empty = mzero
+  (<|>) = mplus
+
+instance MonadPlus UnrenderResult where
+  mzero = Mismatch
+  mplus Mismatch m = m
+  mplus (UnrenderError _) m = m
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply libs/wire-api/src/Wire/API/Routes/MultiVerb.hs:105]: You're right. Good catch!

File: libs/wire-api/src/Wire/API/Routes/MultiVerb.hs [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+  responseRender (AcceptHeader acc) x =
+    M.mapAcceptMedia (map (uncurry mkRenderOutput) (allMimeRender (Proxy @cs) x)) acc
+    where
+      mkRenderOutput :: M.MediaType -> LByteString -> (M.MediaType, RenderOutput)
+      mkRenderOutput c body =
+        (c,) . roAddContentType c $
+          RenderOutput
+            { roStatus = statusVal (Proxy @s),
+              roBody = body,
+              roHeaders = []
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply libs/wire-api/src/Wire/API/Routes/MultiVerb.hs:131]: I think so. A plain response adds no extra headers. When you want headers, you can use a `WithHeaders` wrapper, which takes the `RenderOutput` returned by the underlying `Respond` type and adds headers to the (empty) list.

File: libs/wire-api/src/Wire/API/Routes/MultiVerb.hs [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+  fromHeaders :: Headers hs a -> b
+  toHeaders :: b -> Headers hs a
+
+instance AsHeaders hs a (Headers hs a) where
+  fromHeaders = id
+  toHeaders = id
+
+data DescHeader (name :: Symbol) (desc :: Symbol) (a :: *)
+
+-- convert a list of Header to a list of Servant.Header
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply libs/wire-api/src/Wire/API/Routes/MultiVerb.hs:196]: I know the singular here looks weird, but `Header` and `Headers` are different types, and here I do mean `Header`.

[the remark this answers is not in the document]
Comment (mdimjasevic) [reply libs/wire-api/src/Wire/API/Routes/MultiVerb.hs:197]:
```suggestion
-- convert a list of Header to a list of Servant.Header
++ convert a list of 'Header's to a list of 'Servant.Header's
```

**Review by akshaymankar — COMMENTED** [review]
Overall looks very neat :sparkles: 
I have posted a few questions in comments.

File: libs/wire-api/src/Wire/API/Routes/MultiVerb.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (akshaymankar) [reply libs/wire-api/src/Wire/API/Routes/MultiVerb.hs:132]: Missed the line above.

**Review by akshaymankar — APPROVED** [review]

**Pull request merged**